fem/gridfunc.hpp

```diff
@@ -877,6 +879,164 @@ class ExtrudeCoefficient : public Coefficient
 
+/// Base class for KDTreeNodalProjection.
+class BaseKDTreeNodalProjection
+{
+
+public:
+   virtual ~BaseKDTreeNodalProjection()
+   {}
+
+   /// The projection method can be called as many time as necessary with
+   /// different sets of coordinates and corresponding values. For vector
+   /// grid function, users have to specify the data ordering and for all
+   /// cases the user can modify the error tolerance err to smaller or
+   /// bigger value. A node in the target grid function is matching
+   /// a point with coordinates specified in the vector coords if the
+   /// distance between them is smaller than lerr.
+   virtual
+   void Project(const Vector& coords,const Vector& src,
+                int ordering,double lerr)=0;
+
+   /// The project method can be called as many times as necessary with
+   /// different grid functions gf. A node in the target grid function is
+   /// matching a node from the source grid function if the distance
+   /// between them is smaller than lerr.
+   virtual
+   void Project(const GridFunction& gf, double lerr)=0;
+};
+
+
+
+/// The class provides methods for projecting function values evaluated on a
+/// set of points to a grid function. The values are directly copied to the
+/// nodal values of the target grid function if any of the points is matching
+/// a node of the grid function. For example, if a parallel grid function is
+/// saved in parallel, every saved chunk can be read on every other process
+/// and mapped to a local grid function that does not have the same structure
+/// as the original one. The functionality is based on a kd-tree search in a
+/// cloud of points.
+template<int kdim=3>
+class KDTreeNodalProjection:public BaseKDTreeNodalProjection
+{
+private:
+
+   /// Pointer to the KDTree
+   std::unique_ptr<KDTree<int,double,kdim>> kdt;
+
+   /// Pointer to the target grid function
+   GridFunction* dest;
+
+   /// Upper corner of the bounding box
+   Vector maxbb;
+
+   /// Lower corner of the bounding box
+   Vector minbb;
+
+public:
+   /// The constructor takes as input an L2 or H1 grid function (it can be
+   /// a vector grid function). The Project method copies a set of values
+   /// to the grid function.
+   KDTreeNodalProjection(GridFunction& dest_)
+   {
+      dest=&dest_;
+      FiniteElementSpace* space=dest->FESpace();
+
+      MFEM_VERIFY(
+         dynamic_cast<const H1_FECollection*>(space->FEColl()) != nullptr ||
+         dynamic_cast<const L2_FECollection*>(space->FEColl()) != nullptr,
+         "Error!");
+
+      Mesh* mesh=space->GetMesh();
+
+      const int dim=mesh->SpaceDimension();
+      MFEM_VERIFY(kdim==dim, "GridFunction dimension does not match!");
+
+      kdt=std::unique_ptr<KDTree<int,double,kdim>>(
+             new KDTree<int,double,kdim>());
+
+      std::vector<bool> indt;
+      indt.resize(space->GetVSize()/space->GetVDim(), true);
+
+      minbb.SetSize(dim);
+      maxbb.SetSize(dim);
+
+      //set the loocal coordinates
```

Comment:
Space after `//` (also in other places below)
```suggestion
      // set the loocal coordinates
```

Reply:
done